### PR TITLE
Correct derivation of Penumbra asset ID

### PR DESCRIPTION
**Pull request by hdevalence**
Rather than manually computing the asset ID, we use the existing `penumbra_asset::asset::Metadata` structure to derive it from the metadata. This also ensures that we're running our transported `Metadata` objects through the (de)serialization code we'll eventually use to process them in the client.

### CODE REVIEW

File: tools/compiler/src/processor.rs [hunk 1]
```diff
@@ -60,21 +60,53 @@ pub async fn generate_registry() -> AppResult<()> {
...
+    };
+
+    // Prefix the channel to the base denom.
+    prefix_channel(&mut pb_metadata.base);
+    // Prefix the channel to the display denom.
+    prefix_channel(&mut pb_metadata.display);
+    // Prefix the channel to all denom units.
+    for denom_unit in pb_metadata.denom_units.iter_mut() {
+        prefix_channel(&mut denom_unit.denom);
+    }
```
Comment (hdevalence) [inline tools/compiler/src/processor.rs:81]: We should apply the prefix across all denoms and units. Also, I think the Rust `Metadata` parsing will reject if the base and display don't index into `denom_units`.

**Comment by hdevalence** [comment]
We should check this against the actual IBC data.

File: tools/compiler/src/processor.rs [hunk 2]
```diff
@@ -60,21 +60,53 @@ pub async fn generate_registry() -> AppResult<()> {
...
+            if source_asset.type_asset == AssetTypeAsset::Ics20 {
                 continue;
             }
-            let metadata = convert_to_proto_metadata(&ibc_asset, asset)?;
-            all_metadata.push(metadata);
+            // Turn the asset back into JSON so we can deserialize it as a penumbra Metadata
+            let asset_json = serde_json::to_string(&source_asset)?;
+            let transferred_asset =
+                transport_metadata_along_channel(&ibc_data, serde_json::from_str(&asset_json)?);
+            tracing::info!(?asset_json, transferred_asset_json = ?serde_json::to_string(&transferred_asset));
```
Comment (grod220) [inline tools/compiler/src/processor.rs:108]: Is this outputting something helpful for you? I just see what feels like a bunch of noise. But maybe it's me not familiar with this tracing library.

Screenshot 2024-04-11 at 1 00 45 PM

File: tools/compiler/src/processor.rs [hunk 3]
```diff
@@ -60,21 +60,53 @@ pub async fn generate_registry() -> AppResult<()> {
...
+    for denom_unit in pb_metadata.denom_units.iter_mut() {
+        prefix_channel(&mut denom_unit.denom);
+    }
+
+    // Delete the asset ID, so that it will be recomputed with the adjusted base denom.
+    // Without this, decoding will fail because the asset ID won't match.
+    pb_metadata.penumbra_asset_id = None;
+
+    tracing::debug!(?pb_metadata, "new");
+    Metadata::try_from(pb_metadata).unwrap()
```
Comment (grod220) [inline tools/compiler/src/processor.rs:88]: Think we should return a result here. I'll get this in a later PR though.

File: tools/compiler/src/processor.rs (hunk 1 above)
Comment (grod220) [reply tools/compiler/src/processor.rs:81]: Ah, we seemed to miss this in our current registry too: https://github.com/penumbra-zone/web/blob/main/packages/constants/src/local-asset-registry.json#L176-L188. Good to know.

File: tools/compiler/src/processor.rs [hunk 4]
```diff
@@ -60,21 +60,53 @@ pub async fn generate_registry() -> AppResult<()> {
...
+    tracing::debug!(?pb_metadata, "original");
+
+    let prefix_channel = |x: &mut String| {
+        *x = format!("transfer/{}/{}", ibc_data.ibc_channel, x);
+    };
+
+    // Prefix the channel to the base denom.
+    prefix_channel(&mut pb_metadata.base);
+    // Prefix the channel to the display denom.
+    prefix_channel(&mut pb_metadata.display);
```
Comment (grod220) [inline tools/compiler/src/processor.rs:77]: Do we indeed want to show `transfer/channel-0/usomo` in our UI if we believe it's the canonical one? I believe minifront is using display in a few places 🤔 . I suppose if we rely on symbol, this wouldn't be an issue anyway. Though, a native assets on two different connected chains could in theory conflict on symbol.

File: tools/compiler/src/parser.rs [hunk 5]
```diff
@@ -74,5 +74,6 @@ pub fn get_chain_configs() -> AppResult<Vec<ChainConfig>> {
             };
             Ok(chain_config)
         })
-        .collect()
+        .filter_map(|result| result.ok())
```
Comment (grod220) [inline tools/compiler/src/parser.rs:77]: Don't we want to throw if the user did not set up the file structure correctly? Wouldn't this mean it kinda would silently fail?

File: tools/compiler/src/processor.rs (hunk 2 above)
Comment (hdevalence) [reply tools/compiler/src/processor.rs:108]: Yeah I was using it to inspect the JSON output and determine how the encodings differed during debugging. We can remove

File: tools/compiler/src/processor.rs (hunk 3 above)
Comment (hdevalence) [reply tools/compiler/src/processor.rs:88]: Don't think so, if we had valid Metadata to start, our modifications should also be valid, or else the compiler should fail.

File: tools/compiler/src/processor.rs (hunk 4 above)
Comment (hdevalence) [reply tools/compiler/src/processor.rs:77]: Yeah I think we should use this as a forcing function to only ever use the symbol files in normie UI

File: tools/compiler/src/parser.rs (hunk 5 above)
Comment (hdevalence) [reply tools/compiler/src/parser.rs:77]: Yeah, I did this along the way to starting towards having it consume a git submodule and change the input directory structure, which I abandoned and then wrote up as a follow on issue. Getting output of some kind felt (feels) more important

**Pull request merged**